src/unicode.cpp

```diff
@@ -38,4 +64,54 @@ namespace ada::unicode {
     return c == '\t' || c == '\n' || c == '\r';
   }
 
+  unsigned convert_hex_to_binary(const char c) {
+    if (c >= '0' && c <= '9')
+      return c - '0';
+    if (c >= 'A' && c <= 'F')
+      return 10 + (c - 'A');
+    if (c >= 'a' && c <= 'f')
+      return 10 + (c - 'a');
+  }
+
+  /**
+   * Taken from Node.js
+   * https://github.com/nodejs/node/blob/main/src/node_url.cc#L245
+   *
+   * @see https://encoding.spec.whatwg.org/#utf-8-decode-without-bom
+   */
+  std::string utf8_decode_without_bom(const std::string_view input) {
```

Comment:
If this function is important performance-wise, then we can make it more efficient. We can probably double the efficiency (but it only matters if it is 'hot'). Let us revisit once we have benchmarks.

Reply:
This is one of the hot paths. Let's definitely look into this.